### PR TITLE
Add decrypt/sign and simplify things

### PR DESCRIPTION
**Pull request by hug-dev**
I removed what is not immediatly needed for a demo.

### CODE REVIEW

File: src/subcommands/sign.rs [hunk 1]
```diff
@@ -19,34 +21,63 @@ pub struct Sign {
 
     /// String of UTF-8 text
     input_data: String,
+
+    /// Encode the signature in ASN.1 format (for ECC signatures
+    /// only).
+    #[structopt(long)]
+    encode_asn1: bool,
```
Comment (hug-dev) [inline src/subcommands/sign.rs:28]: I'd be for removing that flag and always doing ASN1 encoding until someone needs it

File: src/subcommands/sign.rs [hunk 2]
```diff
@@ -19,34 +21,63 @@ pub struct Sign {
...
+}
+
+#[derive(Serialize, Deserialize)]
+struct EccSignature {
```
Comment (hug-dev) [inline src/subcommands/sign.rs:32]: Nice 👌

File: src/subcommands/sign.rs [hunk 3]
```diff
@@ -19,34 +21,63 @@ pub struct Sign {
...
                 info!("Signing data...");
-                Ok(basic_client.psa_sign_hash(
-                    self.key_name.clone(),
-                    self.input_data.as_bytes(),
-                    alg,
-                )?)
+                let mut sig = basic_client.psa_sign_hash(self.key_name.clone(), &hash, alg)?;
+                if alg.is_ecc_alg() && self.encode_asn1 {
+                    let mut asn1_signature = vec![0; 1000];
+                    let size = picky_asn1_der::to_bytes(
```
Comment (hug-dev) [inline src/subcommands/sign.rs:59]: Can you use the `picky_asn1_der::to_vec` method to directly generate a vector of the appropriate length without having to truncate a buffer?

File: src/subcommands/sign.rs [hunk 4]
```diff
@@ -19,34 +21,63 @@ pub struct Sign {
...
+                        &EccSignature {
+                            r: IntegerAsn1::from_bytes_be_unsigned(sig[0..sig.len() / 2].to_vec()),
```
Comment (hug-dev) [inline src/subcommands/sign.rs:61]: Maybe [`split_off`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off) can be used well here for that

File: src/subcommands/sign.rs [hunk 5]
```diff
@@ -55,3 +86,18 @@ impl Sign {
...
+        Hash::Sha224 => Box::from(sha2::Sha224::new()),
+        Hash::Sha256 => Box::from(sha2::Sha256::new()),
+        Hash::Sha384 => Box::from(sha2::Sha384::new()),
+        Hash::Sha512 => Box::from(sha2::Sha512::new()),
+        _ => {
+            err!("Hashing algorithm ({:?}) not supported", alg);
+            return Err(ToolErrorKind::NotSupported.into());
+        }
+    };
+    hasher.update(&data);
```
Comment (hug-dev) [inline src/subcommands/sign.rs:101]: Is there a maximal amount of data that one `update` can take at a time? As in, maybe for very large text to sign, we might want to do it 245 bytes per 256 bytes

**Review by hug-dev — COMMENTED** [review]
Thanks!!
Added some comments but we could also address them in subsequent PRs

File: src/subcommands/sign.rs (hunk 5 above)
Comment (ionut-arm) [reply src/subcommands/sign.rs:101]: > Is there a maximal amount of data that one update can take at a time?

I don't think so, it's probably more to allow you to call it multiple times if you need to.

File: src/subcommands/sign.rs [hunk 6]
```diff
@@ -21,32 +23,51 @@ pub struct Sign {
...
-                    alg,
-                )?)
+                let mut sig = basic_client.psa_sign_hash(self.key_name.clone(), &hash, alg)?;
+                if alg.is_ecc_alg() {
+                    let r = IntegerAsn1::from_bytes_be_unsigned(sig.split_off(sig.len() / 2));
+                    sig = picky_asn1_der::to_vec(&EccSignature {
+                        r,
+                        s: IntegerAsn1::from_bytes_be_unsigned(sig),
+                    })
+                    .unwrap();
```
Comment (hug-dev) [inline src/subcommands/sign.rs:58]: I am ok to keep this for now but I think we should add a new error variant and convert from that

**Review by hug-dev — COMMENTED** [review]
I am ok for this as it is now!

File: src/subcommands/sign.rs (hunk 6 above)
Comment (ionut-arm) [reply src/subcommands/sign.rs:58]: Ah, lol, I forgot about that one, will fix in another PR

**Pull request merged**